packages/scrapers-v2/src/runtime/logger.ts

```diff
@@ -0,0 +1,175 @@
+import { Pipeline } from "./types";
+import { ResultType } from "@graduate/common";
+import { FilterError } from "./pipeline";
+import { CatalogEntryType } from "../classify/types";
+
+/**
+ * Logs the progress of the scrape so the developer knows the scraper isn't deadlocked.
+ *
+ * As each individual entry pipeline completes, its status is logged. `.` for
+ * success, `-` for filtered out, or a # representing the stage it errored on.
+ *
+ * @param pipelines The in-progress pipelines
+ */
+export const logProgress = async <T>(
+  pipelines: Array<Promise<Pipeline<T>>>
+) => {
+  // set handlers to log the result of each pipeline
+  for (const promise of pipelines) {
+    promise.then(({ result, trace }) => {
+      if (result.type === ResultType.Ok) {
+        process.stdout.write(".");
+      } else if (result.err[0] instanceof FilterError) {
+        process.stdout.write("-");
+      } else {
+        process.stdout.write(String(trace.length));
+      }
+    });
+  }
+
+  const awaited = await Promise.all(pipelines);
+  process.stdout.write("\n");
+  return awaited;
+};
+
+/**
+ * Logs scrape summary information, error messages, and the URLs of the catalog
+ * entries that errored.
+ *
+ * Note: To display fewer stacktraces, the logger will try to aggregate them.
+ * However, this doesn't always work with async stacktraces, so there might
+ * sometimes appear two of the same stacktrace.
+ *
+ * @param results The completed pipelines
+ */
+export const logResults = <T>(results: Pipeline<T>[]) => {
+  const stats = new StatsLogger();
+
+  for (let i = 0; i < results.length; i += 1) {
+    const { trace, result, id } = results[i];
+    if (result.type === ResultType.Ok) {
+      stats.recordField("status", "ok");
+      continue;
+    } else if (result.err[0] instanceof FilterError) {
+      stats.recordField("status", "filtered");
+      stats.recordField("filtered", result.err[0].actual);
+      if (result.err[0].actual === CatalogEntryType.Unknown) {
+        console.log("entry with unknown type:", id.toString());
+      }
+      continue;
+    }
```

Comment:
i prefer to avoid using else because it makes it less readable (in my opinion) because of the indentation–this is a hot take, so instead i'll just make helper functions